app/src/main/java/org/kiwix/kiwixmobile/localFileTransfer/WifiDirectManager.kt

```diff
@@ -243,8 +246,17 @@ class WifiDirectManager @Inject constructor(
     if (BuildConfig.DEBUG) {
       Log.d(TAG, "Starting handshake")
     }
-    peerGroupHandshakeAsyncTask = PeerGroupHandshakeAsyncTask(this)
-      .also { it.execute() }
+    peerGroupHandshake = PeerGroupHandshake(this)
+      .also {
+        CoroutineScope(Dispatchers.Main).launch {
```

Comment:
Constructing your own scope and doing nothing with it is against the goals of structured concurrency. Ideally we would want to use one of the scopes provided to us, to do so we have to answer the question "how long should these tasks run" or "when should these tasks end" alternatively.

Reply:
>and doing nothing with it is against the goals of structured concurrency

@macgills I don't understand this part. Can you please go to the details?

Reply:
So one of the best things about coroutines is structured concurrency ie to call launch/async you have to have a scope and once you cancel that scope all child jobs get cancelled.
You can create your own scope : `CoroutineScope(Dispatchers.IO)` (also you can pass in job/execptionhandler here with `+`)
However to benefit from this scope being cancelled we need to hang on to it in a variable and call `CoroutineScope.cancel` at the right time, wouldn't it be nice if lovely androidx devs thought of this and just made some nice scopes that cancel at appropriate times and we don't have to think about it too much?
Turns out they did! Activity, Fragment(+fragment views), Viewmodel all have scopes and the work how you think they should.
![image](https://user-images.githubusercontent.com/3358282/98111640-d279ec00-1e98-11eb-90ca-37a3ac73ce75.png)


So back to
>"how long should these tasks run" or "when should these tasks end" alternatively.

When should we cancel this work? Do we just want to send files infinitely?